Automattic/wp-calypso
# Framework: Add lerna to publish @automattic/simple-components (Take Two)

Add `lerna` (a monorepo tool) to Calypso, and use it to publish [`@automattic/simple-components`](https://www.npmjs.com/package/@automattic/simple-components) as a separate npm.

This involves adding a `package.json`, `.gitignore`, and `.npmignore` to `client/components`, and an `alias` field to Calypso's webpack configs.

Since we want Calypso (and possibly other consumers) to continue to use untranspiled code (to benefit from features such as tree-shaking), the npms we produce should contain both transpiled and untranspiled versions of the code. See http://2ality.com/2017/07/npm-packages-via-babel.html, and for more information, http://2ality.com/2017/04/transpiling-dependencies-babel.html and http://2ality.com/2017/06/pkg-esnext.html.

Please note https://github.com/Automattic/wp-calypso/pull/16782#discussion_r131103930!

To test:
* Run `npm run distclean && npm start`
* Verify that Calypso works
* Note that `node_modules/@automattic/simple-components` is a symlink to `client/components/`!
* Also verify that building (`npm run build-docker`) and running (`npm run docker`) the Docker container still works (see PCYsg-5XR-p2).

To publish the npm (only if you know what you're doing! :stuck_out_tongue:):
* Run `npx lerna publish --skip-git` 
  * If you have 2fa (in auth-and-writes mode) enabled for your npm account, you need to prepend `NPM_CONFIG_OTP=<your2FAcode>` to that command -- and be quick before it expires while attempting to publish! Related: https://github.com/lerna/lerna/issues/1091 
* When asked for version, select `Prerelease` for `@automattic/simple-components`!)

Related (but not required for this PR!): #19698.
Take One was #16782.
